runtime/src/accounts_index.rs

```diff
@@ -1194,12 +1195,21 @@ impl<T: 'static + Clone + IsCached + ZeroLamport> AccountsIndex<T> {
                     max_clean_root,
                     account_indexes,
                 );
+                empty_key = slot_list.is_empty();
             });
         }
+        if empty_key {
+            let mut w_maps = self.account_maps.write().unwrap();
+            if let Some(x) = w_maps.get(pubkey) {
+                if x.slot_list.read().unwrap().is_empty() {
+                    w_maps.remove(pubkey);
+                }
+            }
+        }
     }
 
-    pub fn can_purge(max_root: Slot, slot: Slot) -> bool {
-        slot < max_root
+    pub fn can_purge(max_clean_root: Slot, max_present_root: Slot, slot: Slot) -> bool {
+        slot < max_clean_root && slot != max_present_root
```

Comment:
I think for clarifying future readers, I think it would be useful to:

1) Add an assert that `max_cleanable_root_in_slot_list < max_clean_root`
2) Add clarifying comment here  along the lines of:

"If we get a slot update `s  != max_cleanable_root_in_slot_list ` for an account where `s <  max_clean_root`, then we know it's safe to delete because:
a)  If `s < max_cleanable_root_in_slot_list`, then we know the update is outdated by a later rooted update, namely the one in `max_cleanable_root_in_slot_list`
b) If `s > max_cleanable_root_in_slot_list`, then because `s < max_clean_root` and we know there are no roots in the slot list between `max_cleanable_root_in_slot_list` and `max_clean_root`, (otherwise there would be a bigger `max_cleanable_root_in_slot_list`, which is a contradiction), then we know `s` must be an unrooted slot less than `max_clean_root` and thus safe to clean as well.

Reply:
Done! (I used your text pretty much verbatim, hope that's OK!)